---
title:  |
  | **Kinship Density Plots**
  | <span style='font-size: 18pt'>*Example code using KRLA data*</span>
author: "Seth Romero"
date: 4/25/2024
output: 
  html_document:
    theme: flatly
    highlight: tango
---

```{r, echo=FALSE}
library(knitr)
opts_chunk$set(echo = T, message = F, warning = F, fig.width = 10, fig.height = 10, fig.align = 'center')
```

### Necessary packages

```{r}
library(tidyr)
library(dplyr)
library(stringr)
library(ggplot2)
library(patchwork)
library(ggridges)
library(colorspace)
library(SNPRelate)

setwd("~/Documents/GitHub/POMA/")
```

### Calculating kinship

This chunk for calculating kinship is taken straight from Caro's *A. thurberianum* analyses. This will generate a kinship file **per population** for every pairwise comparison of individuals within pop. Uses two different methods (MoM and MLE) for estimation.

A VCF file is only needed starting input.

```{r, eval=F}
# Convert VCF to GDS
file <- snpgdsVCF2GDS('filtering/POMA.bithin.q30.i70.maf02.a90.recode.vcf', "POMA.gds", 
                      method=c("biallelic.only", "copy.num.of.ref"),
                      snpfirstdim=FALSE, compress.annotation="LZMA_RA", compress.geno="",
                      ref.allele=NULL, ignore.chr.prefix="chr", verbose=TRUE)

# Open the GDS file
genofile <- snpgdsOpen("POMA.gds")

# Get all sample IDs
all_samples <- read.gdsn(index.gdsn(genofile, "sample.id"))

# Extract population info
# May need to modified depending on format of pop./ind. ID 
info <- data.frame(
  sample = all_samples,
  species = substr(all_samples, 1, 2),
  population = str_split_i(all_samples, '_', 2)
)


test3 <- info %>%
  group_by(population) %>%
  summarize(n = n())
# Get populations
populations <- unique(info$population)

# Loop over each population
for(pop in populations) {
  
  # Get sample IDs for this population
  sample_ids <- info$sample[info$population == pop]
  
  # MoM IBD calculation
  ibd <- snpgdsIBDMoM(genofile, sample.id=sample_ids, snp.id=NULL, autosome.only=FALSE)
  ibd.coeff1 <- snpgdsIBDSelection(ibd)
  write.csv(ibd.coeff1, paste0("mom_", pop, ".csv"))

  # MLE IBD calculation
  ibd <- snpgdsIBDMLE(genofile, sample.id=sample_ids, snp.id=NULL, autosome.only=FALSE)
  ibd.coeff2 <- snpgdsIBDSelection(ibd)
  write.csv(ibd.coeff2, paste0("mle_", pop, ".csv"))
  
}

# Close the GDS file after processing
snpgdsClose(genofile)
```

### Dataframe setup

Looping through .csv's and combining.

```{r}
# making combined dataframe for MoM estimations
df <- matrix(ncol = 5)
colnames(df) <- c('ID1', 'ID2', 'k0', 'k1', 'kinship')
for(pop in populations) {
  file = paste0('mom_', pop, '.csv')
  popKin = read.csv(file)[,-1]
  df = rbind(df, popKin)
}

momKin <- df[-1,] %>%
  mutate(Project = str_split_i(ID1, '_', 2),
         Pop = str_split_i(ID1, '_', 2))

# making combined dataframe for MLE estimations
df <- matrix(ncol = 6)
colnames(df) <- c('ID1', 'ID2', 'k0', 'k1', 'niter', 'kinship')
for(pop in populations) {
  file = paste0('mle_', pop, '.csv')
  popKin = read.csv(file)[,-1]
  df = rbind(df, popKin)
}

mleKin <- df[-1,] %>%
  mutate(Project = str_split_i(ID1, '_', 2),
         Pop = str_split_i(ID1, '_', 2))
```

### Colors

```{r}
col7 <- c("#56B4E9","#0072B2","#F0E442",
          "#CC79A7","#009E73","#D55E00","#E69F00")

col7_d3 <- darken(col7, amount = 0.3)
```

### Plotting

```{r, fig.height=6, fig.width=12}
mlePlot <- ggplot(data = mleKin, aes(x = kinship, y = Pop, fill = Pop, color = Pop)) +
  geom_vline(xintercept = 0, color = 'black') +
  geom_vline(xintercept = 0.25, color = 'grey25', linetype = 'dashed') +
  geom_vline(xintercept = 0.125, color = 'grey25', linetype = 'dashed') +
  geom_ridgeline(aes(height = after_stat(scaled)),
                 scale = 0.8,
                 stat = 'density',
                 position = position_nudge(y = -0.4),
                 trim = T) +
  geom_jitter(pch = 21, size = 2.5,
              aes(height = NULL),
              position = position_jitter(width = 0.005, height = 0.15),
              color = 'black',
              stroke = 0.25, alpha = 0.8) +
  annotate('text',
           x = 0.135,
           y = 6.7,
           label = substitute(paste(italic('UNCLE / NIECE'))),
           angle = 270,
           size = 4) +
  annotate('text',
           x = 0.135,
           y = 1.1,
           label = substitute(paste(italic('HALF SIBS'))),
           angle = 270,
           size = 4) +
  annotate('text',
           x = 0.26,
           y = 6.2,
           label = substitute(paste(italic('PARENT / OFFSPRING'))),
           angle = 270,
           size = 4) +
  annotate('text',
           x = 0.26,
           y = 1.1,
           label = substitute(paste(italic('FULL SIBS'))),
           angle = 270,
           size = 4) +
  # annotate('text',
  #          x = 0.52,
  #          y = 4,
  #          label = substitute(paste(italic('INDIVIDUAL / SELF'))),
  #          angle = 270,
  #          size = 4.5) +
  # annotate('text',
  #          x = 0.185,
  #          y = 6.9,
  #          label = substitute(paste(bold('N (pairwise) = 555'))),
  #          size = 4) +
  scale_fill_manual(values = col7) +
  #scale_color_manual(values = col7_d3) +
  scale_x_continuous(limit = c(-0.01, 0.265), 
                     breaks = c(0, 0.125, 0.25)) +
  coord_cartesian(ylim = c(0.75, 7.25)) +
  labs(x = 'Kinship Coefficient',
       y = 'Population',
       title = 'Maximum Likelihood Estimation') +
  theme_bw() +
  theme(legend.position = 'none',
        panel.grid = element_blank(),
        plot.title = element_text(size = 16),
        axis.text = element_text(size = 16),
        axis.title.x = element_text(size = 24, face = 'bold', vjust = -1),
        axis.title.y = element_text(size = 24, face = 'bold', vjust = 2),
        panel.border = element_rect(linewidth = 1.5, color = "black"),
        plot.margin = margin(20, 10, 20, 20, 'pt'))

momPlot <- ggplot(data = momKin, aes(x = kinship, y = Pop, fill = Pop, color = Pop)) +
  geom_vline(xintercept = 0, color = 'black') +
  geom_vline(xintercept = 0.25, color = 'grey25', linetype = 'dashed') +
  geom_vline(xintercept = 0.125, color = 'grey25', linetype = 'dashed') +
  geom_ridgeline(aes(height = after_stat(scaled)),
                 scale = 0.8,
                 stat = 'density',
                 position = position_nudge(y = -0.4),
                 trim = T) +
  geom_jitter(pch = 21, size = 2.5,
              aes(height = NULL),
              position = position_jitter(width = 0.005, height = 0.15),
              color = 'black',
              stroke = 0.25, alpha = 0.8) +
  annotate('text',
           x = 0.135,
           y = 6.7,
           label = substitute(paste(italic('UNCLE / NIECE'))),
           angle = 270,
           size = 4) +
  annotate('text',
           x = 0.135,
           y = 1.1,
           label = substitute(paste(italic('HALF SIBS'))),
           angle = 270,
           size = 4) +
  annotate('text',
           x = 0.26,
           y = 6.2,
           label = substitute(paste(italic('PARENT / OFFSPRING'))),
           angle = 270,
           size = 4) +
  annotate('text',
           x = 0.26,
           y = 1.1,
           label = substitute(paste(italic('FULL SIBS'))),
           angle = 270,
           size = 4) +
  # annotate('text',
  #          x = 0.39,
  #          y = 6.9,
  #          label = substitute(paste(bold('N (pairwise) = 555'))),
  #          size = 4) +
  scale_fill_manual(values = col7) +
  #scale_color_manual(values = col7_d3) +
  scale_x_continuous(limit = c(-0.01, 0.265), 
                     breaks = c(0, 0.125, 0.25)) +
  coord_cartesian(ylim = c(0.75, 7.25)) +
  labs(x = 'Kinship Coefficient',
       y = 'Population',
       title = 'Method of Moments') +
  theme_bw() +
  theme(legend.position = 'none',
        panel.grid = element_blank(),
        plot.title = element_text(size = 16),
        axis.text = element_text(size = 16),
        axis.title.x = element_text(size = 24, face = 'bold', vjust = -1),
        axis.title.y = element_blank(),
        panel.border = element_rect(linewidth = 1.5, color = "black"),
        plot.margin = margin(20, 20, 20, 10, 'pt'))


related2panel <- mlePlot + momPlot + 
  plot_layout(ncol = 2)

related2panel


ggsave('POMArelatedness2panel.png', plot = related2panel,
       height = 6, width = 12, dpi = 350)
```